packages/web/src/components/buy-sell-modal/SwapTab.tsx

```diff
@@ -0,0 +1,131 @@
+import { useRef } from 'react'
+
+import { Flex, Skeleton } from '@audius/harmony'
+import { Form, FormikProvider } from 'formik'
+
+import { TokenAmountSection } from './TokenAmountSection'
+import { useTokenSwapForm } from './hooks/useTokenSwapForm'
+import { TokenInfo } from './types'
+
+const messages = {
+  youPay: 'You Pay',
+  youReceive: 'You Receive',
+  placeholder: '0.00'
+}
+
+const TokenSectionSkeleton = ({ title }: { title: string }) => (
+  <Flex direction='column' gap='s'>
+    <Skeleton h='xl' w={title === 'input' ? 'unit20' : 'unit24'} />
+    <Skeleton h='unit14' w='100%' />
+  </Flex>
+)
+
+const ExchangeRateSkeleton = () => (
+  <Flex justifyContent='center' p='s'>
+    <Skeleton w='unit30' h='xl' />
+  </Flex>
+)
+
+const SwapFormSkeleton = () => (
+  <Flex direction='column' gap='l'>
+    <TokenSectionSkeleton title='input' />
+    <TokenSectionSkeleton title='output' />
+  </Flex>
+)
+
+export type SwapTabProps = {
+  inputToken: TokenInfo
+  outputToken: TokenInfo
+  min?: number
+  max?: number
+  balance: {
+    get: () => number | undefined
+    loading: boolean
+    formatError: () => string
+  }
+
+  onTransactionDataChange?: (data: {
+    inputAmount: number
+    outputAmount: number
+    isValid: boolean
+  }) => void
+}
+
+export const SwapTab = ({
+  inputToken,
+  outputToken,
+  min,
+  max,
+  balance,
+  onTransactionDataChange
+}: SwapTabProps) => {
+  const {
+    formik,
+    inputAmount,
+    numericInputAmount,
+    outputAmount,
+    isExchangeRateLoading,
+    isBalanceLoading,
+    availableBalance,
+    currentExchangeRate,
+    handleInputAmountChange,
+    handleMaxClick
+  } = useTokenSwapForm({
+    inputToken,
+    outputToken,
+    min,
+    max,
+    balance,
+    onTransactionDataChange
+  })
+
+  // Track if an exchange rate has ever been successfully fetched
+  const hasRateEverBeenFetched = useRef(false)
```

Comment:
can this just be a constant?

Reply:
we need it to persist between renders to avoid showing the loading skeletons